spmd/compiler/api.py

```diff
@@ -358,10 +392,34 @@ def forward(
         self, *args: Tuple[object], **kwargs: Dict[str, object]
     ) -> object:
         if self._compiled_m is None:
+            input_set = set(args)
+
+            def gather_inputs_for_compilation(inps):
+                compile_inps = [
+                    x
+                    if not isinstance(x, torch.Tensor) or x not in input_set
+                    else DTensor.from_local(x, self._schema.mesh, [Shard(0)])
+                    .redistribute(self._schema.mesh, [Replicate()])
+                    .to_local()
+                    for x in inps
+                ]
+                return compile_inps
+
             self._compiled_m = aot_module(
                 self._local_module,
                 partial(self._compile, TrainingPhase.FORWARD),
                 partial(self._compile, TrainingPhase.BACKWARD),
+                pre_compile_fn=gather_inputs_for_compilation,
```

Comment:
oh so this is where we need to patch the aot_function?

Reply:
yeah i added a comment explaining -- the issue is that we want to present "full" logical inputs at capture time, but then we want to pass shraded input at run-time (because we transformed the graph in between).